### ISSUE TITLE
Fix service graph to handle zipkin and jaeger styled spans with mutual exclusion 

### ISSUE THREAD

**Issue by ashishagg** [ExpediaDotCom/haystack#555]
Today, service graph implementation gets messed up when it receives spans in both following styles at the same time. 

1. Zipkin Styled Spans: The server and client spans have the same TraceId, SpanId and ParentSpanId, but 'span.kind' tag is different identifying it is 'server' or 'client'
2. Jaeger Styled Spans: Jaegar produces two spans for every rpc call, with a parent-child relationship. In other words, server span is a child of its client span.  
For reference check [here](https://github.com/opentracing/opentracing.io/issues/102)

Proposal:
All of haystack components can handle zipkin or jaeger styled spans individually but not in a mixed fashion. Adding a logic to handle these styles in one service-mesh deployment adds unncessary complexity. 
Hence introduce a 'mode' in haystack that makes it work either for zipkin or jaeger styled spans.

**Comment by mchandramouli** [comment]
@ashishagg Do we have a 'mode' at the cluster level so all components can behave accordingly. Thinking about it, at present, it will be only traces and graph that care about this relationship.  I agree, it will make the code cleaner to handle one of the modes.

**Comment by mchandramouli** [comment]
@ashishagg My 2 cents: lets make Zipkin as default

**Comment by adriancole** [comment]
I wouldn't call this "Jaeger Styled Spans" as that's a bit narrow. Also, I suspect jaeger already need to support ID sharing as their old instrumentation did this (as they used to be zipkin) also they pitch themselves as a drop-in replacement. Even in B3 itself, we describe the pattern of ignoring the X-B3-ParentSpanId https://github.com/openzipkin/b3-propagation#why-is-parentspanid-propagated

Also, zipkin instrumentation, we have a setting to create both styles of span data. Ex. a tracer setting `supportsJoin` says whether or not it is ok for the server to share a span ID. This was originally created in the zipkin-go-opentracing project, but it is available in a lot of tools, like brave and zipkin-go and zipkin-php. That's because most cloud services (stackdriver, amazon, etc) don't support sharing the same ID.

All that said, many instrumentation are created internal to companies and they look at B3 and implement ID sharing accordingly, even if B3 is not limited to ID sharing style (again the server decides whether or not.. the client can't control this)

Context Summary

This is much more than a Jaeger concern, just maybe just first noticed there. There is instrumentation besides what zipkin produce (hundreds of zipkin libraries folks make internally or publish on their own), likely most choose to share IDs, but you can't control all of them. I'd suggest making it possible to support both patterns in service graph.

Action to take:

In Zipkin's dependency link generator, we support both styles knowing we can't control it anyway. It means some heuristics are baked in. https://github.com/openzipkin/zipkin/blob/master/zipkin/src/main/java/zipkin2/internal/DependencyLinker.java#L181

Whether or not you can reuse this approach depends on if you can see the whole trace tree when you are linking.

**Comment by mchandramouli** [comment]
You nailed it in the last line. Haystack's dependency linker (SpanAccumulator.scala in service-graph) doesn't see the whole trace, its an independent component that reads spans from the main topic and waits for spans with shared IDs and produces the edge in service graph.   So when we get mixed mode,  where client is Zipkin style and the server span uses a new span id, it sort of gets confused. @ashishagg and I were looking at options. Maybe we will give a shot again and see if we can link them correctly.

**Comment by ashishagg** [comment]
Right, we have considered an option to read all spans of a trace in one shot. This will help to deduce the correct edge for a mixed span-styling setup.

Haystack's trace-indexer emits buffered spans grouped by a TraceId every X sec. The graph-builder component can consume from this topic, but here are few things to consider:

a) This span-buffer may not contain all spans for a traceId, indexer can evict span-buffer when its memory limits are hit (spikes in traffic can trigger it).  
b) It also doubles the I/O on kafka brokers,  essentially all spans that indexer reads is output back to kafka in similar form.
c) Lastly, we see a possible case of 3 services, A -> B -> C, all with zipkin styled span. And if B's span goes missing then we may end up creating an edge between A -> C (considering them to be jaeger styled).    

We propose a different approach where graph builder should have the capability to correct itself.

Lets start with a Map<ServiceName, Style>
Style can be `SINGULAR`(zipkin) or `DUAL`(jaeger).

The logic is simple. Whenever we detect an edge, we generally know what 'style' generates it. Now,
a) Look for *server span* serviceName in the map, and if style matches with its previous observation, then produce an edge, else discard.
b) If map doesn't contain a style for 'server span' service, then update its style and produce the edge.
c) Override the style in the map, iff you detect a SINGULAR and previously observed style is DUAL.

Option c) provides autocorrection, it solves many cases including the one we described above. We may generate a wrong edge for first few minutes but then it auto adjusts.

This map is persisted in kafka to handle redeployments or other failures.

**Comment by adriancole** [comment]
PS on this point. I was told that when dapper transitioned, they used the
span kind field in consideration when shared ids are in use. In zipkin of
course we have a shared field that tells you this. Ex when you get a span
with shared value set you don't even have to consider the value of kind.
The logical outcome of how the data is shaped might end up similar. It
would be nice to see a model with test data to prove the point.

>

**Comment by adriancole** [comment]
can someone point me to the code in question by the way? I'd like to see the feed of fields, for example you can aggregate online when you know authoritative remote (calling) service name regardless of parent ID. For example, that's why we've designed brave to have alternate finished span handlers. I think after I dig into this with real data I'll be able to help more?

**Comment by mchandramouli** [comment]
@ashishagg figured a way to deduce the span type of the server (single span id between client and server  vs  dual span id). We are testing it. If it works, we will release that and close this issue. Haystack can support both span types.

**Comment by ashishagg** [comment]
We have added a change in the service-graph implementation that takes care of system that emits both shared(singular) and non-shared(dual) rpc span.
We have deployed this change in our production environment and it has been running good for last 4 days or so. I have detailed out earlier how this works.

https://github.com/ExpediaDotCom/haystack-service-graph/commit/7baeb1a0e23bd613a2d55e5df5b720f85e84fb19

I will let this Issue remain open for next few days to give few more days to run.

**Comment by adriancole** [comment]
Thanks tons.. didn't see that repo somehow until now. watched and starred!

**Comment by mchandramouli** [comment]
Changes to support both styles of spans https://github.com/ExpediaDotCom/haystack-service-graph/pull/65
